lib/scholar/metrics/distance.ex

```diff
@@ -240,25 +368,53 @@ defmodule Scholar.Metrics.Distance do
       iex> y = Nx.tensor([1, 2, 3])
       iex> Scholar.Metrics.Distance.cosine(x, y)
       ** (ArgumentError) expected input shapes to be equal, got {2} != {3}
+
+      iex> x = Nx.tensor([[1, 2, 3], [0, 0, 0], [5, 2, 4]])
+      iex> y = Nx.tensor([[1, 5, 2], [2, 4, 1], [0, 0, 0]])
+      iex> Scholar.Metrics.Distance.cosine(x, y, axes: [1])
+      #Nx.Tensor<
+        f32[3][3]
+        [
+          [0.1704850196838379, 0.2418246865272522, 1.0],
+          [1.0, 1.0, 0.0],
+          [0.3740193247795105, 0.2843400239944458, 1.0]
+        ]
+      >
   """
   @spec cosine(Nx.t(), Nx.t()) :: Nx.t()
-  defn cosine(x, y) do
+  defn cosine(x, y, opts \\ []) do
+    cutoff = 10 * 2.220446049250313e-16
     assert_same_shape!(x, y)
-    norm_x = Nx.LinAlg.norm(x)
-    norm_y = Nx.LinAlg.norm(y)
 
-    cond do
-      norm_x == 0.0 and norm_y == 0.0 ->
-        0.0
+    x = if Nx.rank(x) == 1, do: Nx.new_axis(x, 0), else: x
+    y = if Nx.rank(y) == 1, do: Nx.new_axis(y, 0), else: y
 
-      norm_x == 0.0 or norm_y == 0.0 ->
-        1.0
+    {m, n} = Nx.shape(x)
 
-      true ->
-        numerator = Nx.dot(x, y)
-        denominator = norm_x * norm_y
-        1.0 - numerator / denominator
-    end
+    opts =
+      keyword!(
+        opts,
+        axes: nil
+      )
+
+    norm_x = Nx.LinAlg.norm(x, axes: opts[:axes]) |> Nx.reshape({m, 1}) |> Nx.broadcast({m, n})
+    norm_y = Nx.LinAlg.norm(y, axes: opts[:axes]) |> Nx.reshape({1, m}) |> Nx.broadcast({n, m})
+
+    zero_mask = norm_x == 0.0 and norm_y == 0.0
+    zero_xor_one_mask = Nx.logical_xor(norm_x == 0.0, norm_y == 0.0)
+
+    norm_y = Nx.transpose(norm_y)
+
+    norm_x = Nx.select(Nx.greater(norm_x, cutoff), norm_x, 1.0)
+    norm_y = Nx.select(Nx.greater(norm_y, cutoff), norm_y, 1.0)
+
+    norm_x = x / norm_x
+    norm_y = y / norm_y
+
+    res = Nx.dot(norm_x, Nx.transpose(norm_y))
+    res = Nx.select(zero_xor_one_mask, 0.0, res)
+    res = 1.0 - Nx.select(zero_mask, 1.0, res)
+    if m != 1, do: res, else: res[0][0]
```

Comment:
Why `res[0][0]`? What happens if the input has rank > 2

Reply:
At this moment the algorithm deals only with 2d inputs because it calculate pair-wise distance, but I'll change that to be consistent with the rest of API